sdk/eventhub/azure-eventhubs/azure/eventhub/aio/_producer_client_async.py

```diff
@@ -56,31 +64,71 @@ def __init__(self, host, event_hub_path, credential, **kwargs) -> None:
+        self._lock = asyncio.Lock()  # sync the creation of self._producers
         self._max_message_size_on_link = 0
-
-    async def _init_locks_for_producers(self):
-        if not self._producers:
-            async with self._client_lock:
-                if not self._producers:
-                    num_of_producers = len(await self.get_partition_ids()) + 1
-                    self._producers = [None] * num_of_producers
-                    for _ in range(num_of_producers):
-                        self._producers_locks.append(asyncio.Lock())
-                        # self._producers_locks = [asyncio.Lock()] * num_of_producers
+        self._partition_ids = None
+
+    async def _get_partitions(self):
+        if not self._partition_ids:
+            self._partition_ids = await self.get_partition_ids()
+            for p_id in self._partition_ids:
+                self._producers[p_id] = None
+
+    async def _get_max_mesage_size(self):
+        # pylint: disable=protected-access
+        async with self._lock:
+            if not self._max_message_size_on_link:
+                await self._producers[ALL_PARTITIONS]._open_with_retry()
+                self._max_message_size_on_link = \
+                    self._producers[ALL_PARTITIONS]._handler.message_handler._link.peer_max_message_size \
+                    or constants.MAX_MESSAGE_LENGTH_BYTES
+
+    async def _start_producer(self, partition_id, send_timeout):
+        async with self._lock:
+            await self._get_partitions()
+            if partition_id not in self._partition_ids and partition_id != ALL_PARTITIONS:
+                raise ConnectError("Invalid partition {} for the event hub {}".format(partition_id, self.eh_name))
+
+            if not self._producers[partition_id] or self._producers[partition_id].closed:
+                self._producers[partition_id] = self._create_producer(
+                    partition_id=partition_id,
+                    send_timeout=send_timeout
+                )
+
+    def _create_producer(
+            self, *,
+            partition_id: str = None,
+            send_timeout: float = None,
+            loop: asyncio.AbstractEventLoop = None
+    ) -> EventHubProducer:
+        target = "amqps://{}{}".format(self._address.hostname, self._address.path)
+        send_timeout = self._config.send_timeout if send_timeout is None else send_timeout
+
+        handler = EventHubProducer(
+            self, target, partition=partition_id, send_timeout=send_timeout, loop=loop)
+        return handler
 
     @classmethod
-    def from_connection_string(cls, conn_str, **kwargs):
+    def from_connection_string(
+            cls, conn_str: str,
+            *,
+            event_hub_path: str = None,
+            logging_enable: bool = False,
+            http_proxy: dict = None,
+            auth_timeout: float = 60,
+            user_agent: str = None,
+            retry_total: int = 3,
+            transport_type=None):
```

Comment:
no additional kwargs possible? they are all here?

Reply:
I think they are all here, but will add kwargs for future option support